In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import statsmodels.api as sm
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge, Lasso
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import LabelEncoder, StandardScaler, PolynomialFeatures
from sklearn.metrics import make_scorer, roc_auc_score
from sklearn.metrics import classification_report, accuracy_score
from sklearn.neural_network import MLPClassifier
from skopt import BayesSearchCV
from skopt.space import Real, Categorical, Integer
from sklearn.metrics import accuracy_score, classification_report
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.gaussian_process.kernels import RBF
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.model_selection import StratifiedKFold
from sklearn.gaussian_process import GaussianProcessRegressor as GPR
from sklearn.gaussian_process.kernels import RBF
import warnings
warnings.filterwarnings("ignore")
from scipy import stats

1.   Tomar Advertising.csv y hacer una regresión (linear regression) como benchmark(r^2)
2.   Usar un decision tree y ajustar con dataset (r^2)
3.   Comparar R^2. Escribe observaciones
4.   Repetir (pasos del 1-3) con train test split (random_state=42) (R^2 train, test)
5.   Repetir (pasos del 1-3) con polynomialfeatures(degree=2)

In [2]:
from sklearn.tree import DecisionTreeRegressor 

In [3]:
df = pd.read_csv("Advertising.csv")
df.head()

Unnamed: 0     TV  radio  newspaper  sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

In [4]:
X = df.drop(columns=["sales"])
y = df["sales"]

### 1.

In [5]:
lr=LinearRegression()
lr.fit(X,y)

y_pred = lr.predict(X)

print("R2:", r2_score(y, y_pred))

R2: 0.8972508370448042


### 2.

In [6]:
tree = DecisionTreeRegressor(random_state=42)

tree.fit(X, y)

y_pred_tree = tree.predict(X)

r2_tree = r2_score(y, y_pred_tree)
print("R2 (Decision Tree):", r2_tree)

R2 (Decision Tree): 1.0


### 3.

La regresión lineal explica el 89.7% de las ventas, mostrando buen ajuste y capacidad de generalización. El árbol de decisión tiene un R2 de 1, lo que indica un ajuste perfecto, pero probablemente refleja sobreajuste, por lo que podría no generalizar bien a nuevos datos.

### 4.

In [7]:
Xtrain, Xtest, ytrain, ytest=train_test_split(X,y,test_size=.3,random_state=137)

In [8]:
lr=LinearRegression()
lr.fit(Xtrain,ytrain)

y_pred_train = lr.predict(Xtrain)
y_pred_test = lr.predict(Xtest)

print("R2 train:", r2_score(ytrain, y_pred_train))
print("R2 test:",  r2_score(ytest, y_pred_test))

R2 train: 0.9024054991708091
R2 test: 0.8781075090355257


In [9]:
tree = DecisionTreeRegressor(random_state=42)

tree.fit(Xtrain, ytrain)

y_pred_train = tree.predict(Xtrain)
y_pred_test = tree.predict(Xtest)

print("R2 train:", r2_score(ytrain, y_pred_train))
print("R2 test:",  r2_score(ytest, y_pred_test))

R2 train: 1.0
R2 test: 0.9202993529726364


El modelo de regresión lineal logra un buen ajuste, explicando alrededor del 90% de los datos de entrenamiento y 87.8% de los de prueba. Esto indica que generaliza bien y mantiene un rendimiento estable.

El árbol de decisión tiene un R2 muy similar, aunque la regresión lineal muestra un equilibrio más consistente entre ajuste y generalización.

### 5.

In [10]:
poly = PolynomialFeatures(degree=2)

Xtrain_poly = poly.fit_transform(Xtrain)
Xtest_poly = poly.transform(Xtest)

lr = LinearRegression()
lr.fit(Xtrain_poly, ytrain)

y_pred_train = lr.predict(Xtrain_poly)
y_pred_test = lr.predict(Xtest_poly)

print("R2 train:", r2_score(ytrain, y_pred_train))
print("R2 test:",  r2_score(ytest, y_pred_test))


R2 train: 0.9897399501495978
R2 test: 0.9764748289712576


In [11]:
poly = PolynomialFeatures(degree=2)

Xtrain_poly = poly.fit_transform(Xtrain)
Xtest_poly = poly.transform(Xtest)

tree = DecisionTreeRegressor(random_state=42)
tree.fit(Xtrain_poly, ytrain)

y_pred_train = tree.predict(Xtrain_poly)
y_pred_test = tree.predict(Xtest_poly)

print("R2 train:", r2_score(ytrain, y_pred_train))
print("R2 test:",  r2_score(ytest, y_pred_test))

R2 train: 1.0
R2 test: 0.9813171744678062


La regresión polinómica logra un muy buen ajuste y generaliza bien (R2 train: 0.99, test: 0.98).
El árbol con polinomios tiene un ajuste perfecto en entrenamiento (R2=1.0) y un test ligeramente mejor, pero muestra signos de sobreajuste.